File: tools/canoe-host/canoelib/ui.py
"""Operator-facing output, and the boundary that every entry point runs behind.

The shapes here are contracts, not decoration: `[*] ` opens a step, four
leading spaces mark a detail under it, `    WARNING: ` is a non-fatal problem
and `<prog>: error: ` on stderr is a fatal one. The wiki, the packaged flashers
and the tests all match on them.
"""

from __future__ import annotations

import sys
from collections.abc import Callable, Sequence

from .errors import CanoeError


def step(text: str) -> None:
    """Announce a phase of the run."""
    print(f"\n[*] {text}", flush=True)


def note(text: str) -> None:
    """Report a detail underneath the current step."""
    print(f"    {text}", flush=True)


def warn(text: str) -> None:
    """Report a problem the operator must know about but that is not fatal."""
    print(f"    WARNING: {text}", file=sys.stderr, flush=True)


def emit(text: str) -> None:
    """Write report text verbatim, for the closing summaries."""
    print(text, flush=True)

def ask_choice(question: str, choices: Sequence[str], default: str | None = None) -> str:
    """Ask for one value from a finite operator-facing choice list."""
    options = "/".join(choices)
    suffix = f" [{default}]" if default is not None else ""
    while True:
        answer = input(f"{question} ({options}){suffix}: ").strip().lower()
        selected = answer or (default or "")
        if selected in choices:
            return selected
        warn(f"choose one of: {options}")


def ask_yes_no(question: str, default: bool = True) -> bool:
    """Ask a yes/no question with a deterministic default."""
    suffix = "[Y/n]" if default else "[y/N]"
    while True:
        answer = input(f"{question} {suffix}: ").strip().lower()
        if not answer:
            return default
        if answer in ("y", "yes"):
            return True
        if answer in ("n", "no"):
            return False
        warn("answer yes or no")


def run_entry(prog: str, run: Callable[[Sequence[str]], None], argv: Sequence[str]) -> int:
    """Run a tool body and map its outcome onto the documented exit contract."""
    try:
        run(argv)
    except CanoeError as exc:
        print(f"{prog}: error: {exc}", file=sys.stderr, flush=True)
        return 1
    except KeyboardInterrupt:
        print(f"{prog}: interrupted", file=sys.stderr, flush=True)
        return 130
    return 0

File: tools/canoe/src/ui.rs
use std::io::{self, BufRead, Write};

use crate::error::CanoeError;

pub fn step(text: &str) {
    println!("\n[*] {text}");
    let _ = io::stdout().flush();
}

pub fn note_to<W: Write + ?Sized>(writer: &mut W, text: &str) -> Result<(), CanoeError> {
    writeln!(writer, "    {text}")?;
    writer.flush()?;
    Ok(())
}

pub fn note(text: &str) {
    let mut stdout = io::stdout().lock();
    let _ = note_to(&mut stdout, text);
}

pub fn warn(text: &str) {
    eprintln!("    WARNING: {text}");
    let _ = io::stderr().flush();
}

pub fn emit(text: &str) {
    println!("{text}");
    let _ = io::stdout().flush();
}

fn read_answer<R: BufRead + ?Sized, W: Write + ?Sized>(
    reader: &mut R,
    writer: &mut W,
    question: &str,
) -> Result<String, CanoeError> {
    write!(writer, "{question}")?;
    writer.flush()?;
    let mut answer = String::new();
    let read = reader.read_line(&mut answer)?;
    if read == 0 {
        return Err(CanoeError::message("input closed"));
    }
    Ok(answer.trim().to_ascii_lowercase())
}


pub fn ask_choice<R: BufRead + ?Sized, W: Write + ?Sized>(
    reader: &mut R,
    writer: &mut W,
    question: &str,
    choices: &[&str],
    default: Option<&str>,
) -> Result<String, CanoeError> {
    let options = choices.join("/");
    let suffix = default.map_or_else(String::new, |value| format!(" [{value}]"));
    loop {
        let answer = read_answer(reader, writer, &format!("{question} ({options}){suffix}: "))?;
        let selected = if answer.is_empty() {
            default.unwrap_or_default().to_owned()
        } else {
            answer
        };
        if choices.iter().any(|choice| *choice == selected) {
            return Ok(selected);
        }
        warn(&format!("choose one of: {options}"));
    }
}

pub fn ask_yes_no(question: &str, default: bool) -> Result<bool, CanoeError> {
    let stdin = io::stdin();
    let stdout = io::stdout();
    ask_yes_no_from(&mut stdin.lock(), &mut stdout.lock(), question, default)
}

pub fn ask_yes_no_from<R: BufRead + ?Sized, W: Write + ?Sized>(
    reader: &mut R,
    writer: &mut W,
    question: &str,
    default: bool,
) -> Result<bool, CanoeError> {
    let suffix = if default { "[Y/n]" } else { "[y/N]" };
    loop {
        let answer = read_answer(reader, writer, &format!("{question} {suffix}: "))?;
        if answer.is_empty() {
            return Ok(default);
        }
        match answer.as_str() {
            "y" | "yes" => return Ok(true),
            "n" | "no" => return Ok(false),
            _ => warn("answer yes or no"),
        }
    }
}

pub fn run_entry<F>(prog: &str, run: F, argv: &[String]) -> i32
where
    F: FnOnce(&[String]) -> Result<(), CanoeError>,
{
    match run(argv) {
        Ok(()) => 0,
        Err(CanoeError::Message(message)) => {
            eprintln!("{prog}: error: {message}");
            1
        }
        Err(CanoeError::Io(error)) => {
            eprintln!("{prog}: error: {error}");
            1
        }
    }
}
